### PR TITLE
Add CLI support for branch in direct GitHub URL

### PR DESCRIPTION
**Pull request by fdel-car**
### What does it do?

It modifies the 2 following packages:

- `create-strapi-starter`: Now also use `filepath` from `github-url-parse` package since a branch name could contain a `/`. When it does, `ref` does not contain the full branch name and `filepath` will need to be appended to it.
- `strapi-generate-new`:  Switch `hosted-git-info` to `github-url-parse` in order to reproduce the behavior in the package above.

### Why is it needed?

It allows users to specify a branch when they use the full GitHub URL in the CLI associated with those packages.

### How to test it?

You can directly use a command like `./node_modules/.bin/create-strapi-starter` and specify a branch in the starter GitHub URL (e.g: https://github.com/strapi/strapi-starter-next-corporate/tree/feat/i18n).

### CODE REVIEW

**Comment by markkaylor** [comment]
Hey @fdel-car, I can't remember, did we have a reason for keeping the url parser?  It seems like we could just fetch everything from GitHub.

**Comment by fdel-car** [comment]
Hmm, we'll still need to isolate the full name (`user/name`) of the repository - which is certainly not hard but already done by the package, among other things - in order to fetch the required information from the GitHub API, then we'll need to find the branch name in the given URL ourself.

If we want to keep the parsing behavior simple, what we could do is to let the user choose the branch among the list of branches returned by GitHub (by the branches endpoint, e.g: https://api.github.com/repos/strapi/strapi-starter-next-corporate/branches) through input with auto-completion once fetched, for instance, it's probably overkill for such a feature. This behavior should be opt-in with a flag and just use the default branch by default.

There is a 60 requests per minute limit for unauthenticated GitHub API calls, it shouldn't matter but I just wanted to let you know.

File: packages/create-strapi-starter/utils/fetch-github.js [hunk 1]
```diff
@@ -45,32 +47,38 @@ async function getDefaultBranch(repo) {
...
  * @param  {string} tmpDir Path to temporary directory
  */
-async function downloadGithubRepo(starterUrl, tmpDir) {
-  const { full_name, ref, usedShortcut } = await getRepoInfo(starterUrl);
+async function downloadGitHubRepo(starterUrl, tmpDir) {
+  const { full_name, ref, filepath, usedShortcut } = await getRepoInfo(starterUrl);
   const default_branch = await getDefaultBranch(full_name);
 
-  const branch = ref ? ref : default_branch;
+  let branch = default_branch;
```
Comment (alexandrebodin) [inline packages/create-strapi-starter/utils/fetch-github.js:76]: if we return branch from getRepoInfo this makes this code way simpler

Comment (fdel-car) [reply packages/create-strapi-starter/utils/fetch-github.js]: Done in both packages :)
I took the opportunity to change how we parse the shorthands a bit to unify the return values, I hope that's ok

File: packages/strapi-generate-new/lib/utils/merge-template.js [hunk 2]
```diff
@@ -169,36 +171,66 @@ async function checkTemplateContentsStructure(templateContentsPath) {
...
-        project: `strapi-template-${project}`,
-      };
-    }
-  } catch (error) {
-    // If it's not a GitHub URL, then assume it's a shorthand for an official template
-    return {
-      user: 'strapi',
-      project: `strapi-template-${template}`,
-    };
+function parseShorthand(template) {
```
Comment (alexandrebodin) [inline packages/strapi-generate-new/lib/utils/merge-template.js:174]: can you move the logic into a fetch-github file like in the other CLI ?

Comment (fdel-car) [reply packages/strapi-generate-new/lib/utils/merge-template.js]: Sure, it's done

File: packages/strapi-generate-new/lib/utils/fetch-github.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  const name = `strapi-template-${template}`;
+  return {
+    name,
+    fullName: `strapi/${name}`,
+  };
+}
+
+async function getRepoInfo(template) {
```
Comment (alexandrebodin) [inline packages/strapi-generate-new/lib/utils/fetch-github.js:28]: Can you add the jsdoc on the functions :)

**Review by remidej — COMMENTED** [review]
I tested the create-strapi-app with a [custom template branch](https://github.com/strapi/strapi-template-corporate/tree/feat/i18n), and everything worked fine 👌 

This will be very helpful!

**Review by alexandrebodin — APPROVED** [review]
LGTM. Nice job 👍

**Pull request merged**